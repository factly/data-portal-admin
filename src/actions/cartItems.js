import axios from '../utils/axios';
import {
  ADD_CARTITEMS,
  SET_CARTITEM_LOADING,
  SET_CARTITEM_REQUEST,
  SET_CARTITEM_IDS,
  CARTITEM_API,
} from '../constants/cartItems';
import { addProducts } from './products';
import { getIds, getValues, deleteKeys, buildObjectOfItems } from '../utils/objects';

export const loadCartItems = (id, page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      cartItems: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setCartItemIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CARTITEM_API(id)}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { id, page: page, limit: limit, ids: currentPageIds };
    dispatch(setCartItemRequest(currentReq, total));
    dispatch(addCartItems(nodes));
    dispatch(setCartItemIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

const setLoading = (loading) => {
  return {
    type: SET_CARTITEM_LOADING,
    payload: { loading },
  };
};

export const addCartItems = (cartItems) => (dispatch) => {
  const products = getValues(cartItems, 'product');
  dispatch(addProducts(products));

  dispatch({
    type: ADD_CARTITEMS,
    payload: {
      cartItems: buildObjectOfItems(deleteKeys(cartItems, ['product'])),
    },
  });
};

const setCartItemRequest = (req, total) => {
  return {
    type: SET_CARTITEM_REQUEST,
    payload: { req, total },
  };
};

const setCartItemIds = (ids) => {
  return {
    type: SET_CARTITEM_IDS,
    payload: { ids },
  };
};
