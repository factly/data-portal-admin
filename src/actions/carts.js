import axios from '../utils/axios';
import {
  baseUrl,
  LOADING_CARTS,
  LOADING_CART_DETAILS,
  LOAD_CARTS_SUCCESS,
  LOAD_CARTS_FAILURE,
  GET_CART_ITEMS_SUCCESS,
  GET_CART_ITEMS_FAILURE,
} from '../constants/carts';

export const loadCarts = (page, limit) => {
  return async (dispatch, getState) => {
    let url = baseUrl;
    if (page && limit) {
      url = `${url}?page=${page}&limit=${limit}`;
    }

    dispatch(loadingCarts());

    const response = await axios({
      url: url,
      method: 'get',
    }).catch((error) => {
      dispatch(loadCartsFailure(error.message));
    });

    if (response) {
      dispatch(loadCartsSuccess(response.data));
    }
  };
};

export const getCartItems = (id, page, limit) => {
  return async (dispatch, getState) => {
    let url = `${baseUrl}/${id}/items`;
    if (page && limit) {
      url = `${url}?page=${page}&limit=${limit}`;
    }

    dispatch(loadingCartDetails());

    const response = await axios({
      url: url,
      method: 'get',
    }).catch((error) => {
      dispatch(getCartItemsFailure(error.message));
    });

    if (response) {
      dispatch(getCartItemsSuccess(response.data));
    }
  };
};

const loadingCarts = () => {
  return {
    type: LOADING_CARTS,
  };
};

const loadingCartDetails = () => {
  return {
    type: LOADING_CART_DETAILS,
  };
};

const loadCartsSuccess = (data) => {
  return {
    type: LOAD_CARTS_SUCCESS,
    payload: {
      items: data.nodes,
      total: data.total,
    },
  };
};

const loadCartsFailure = (message) => {
  return {
    type: LOAD_CARTS_FAILURE,
    payload: message,
  };
};

const getCartItemsSuccess = (data) => {
  return {
    type: GET_CART_ITEMS_SUCCESS,
    payload: {
      items: data.nodes,
      total: data.total,
    },
  };
};

const getCartItemsFailure = (message) => {
  return {
    type: GET_CART_ITEMS_FAILURE,
    payload: message,
  };
};
