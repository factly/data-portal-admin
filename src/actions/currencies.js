import axios from '../utils/axios';
import {
  ADD_CURRENCY,
  ADD_CURRENCIES,
  SET_CURRENCY_LOADING,
  SET_CURRENCY_REQUEST,
  SET_CURRENCY_IDS,
  RESET_CURRENCY,
  CURRENCY_API,
} from '../constants/currencies';
import { getIds, buildObjectOfItems } from '../utils/objects';

export const loadCurrencies = (page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      currencies: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setCurrencyIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CURRENCY_API}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { page: page, limit: limit, ids: currentPageIds };
    dispatch(setCurrencyRequest(currentReq, total));
    dispatch(addCurrencies(nodes));
    dispatch(setCurrencyIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

export const createCurrency = (data) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    await axios({
      url: CURRENCY_API,
      method: 'post',
      data: data,
    });

    dispatch(resetCurrency());
    dispatch(setLoading(false));
  };
};

export const updateCurrency = (id, data) => {
  return async (dispatch, getState) => {
    let url = `${CURRENCY_API}/${id}`;

    dispatch(setLoading(true));

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    });

    dispatch(addCurrency(response.data));
    dispatch(setLoading(false));
  };
};

export const deleteCurrency = (id) => {
  return async (dispatch, getState) => {
    let url = `${CURRENCY_API}/${id}`;

    dispatch(setLoading(true));

    await axios({
      url: url,
      method: 'delete',
    });

    dispatch(resetCurrency());
    dispatch(setLoading(false));
  };
};

export const getCurrency = (id) => {
  return async (dispatch, getState) => {
    const {
      currencies: { items },
    } = getState();

    if (items[id]) {
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CURRENCY_API}/${id}`,
      method: 'get',
    });
    dispatch(addCurrency(response.data));

    dispatch(setLoading(false));
  };
};

const setLoading = (loading) => {
  return {
    type: SET_CURRENCY_LOADING,
    payload: { loading },
  };
};

const addCurrency = (currency) => {
  return {
    type: ADD_CURRENCY,
    payload: { currency },
  };
};

export const addCurrencies = (currencies) => {
  return {
    type: ADD_CURRENCIES,
    payload: {
      currencies: buildObjectOfItems(currencies),
    },
  };
};

const setCurrencyRequest = (req, total) => {
  return {
    type: SET_CURRENCY_REQUEST,
    payload: { req, total },
  };
};

const setCurrencyIds = (ids) => {
  return {
    type: SET_CURRENCY_IDS,
    payload: { ids },
  };
};

const resetCurrency = () => {
  return {
    type: RESET_CURRENCY,
  };
};
