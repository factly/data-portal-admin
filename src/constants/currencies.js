export const ADD_CURRENCY = 'ADD_CURRENCY';
export const ADD_CURRENCIES = 'ADD_CURRENCIES';
export const SET_CURRENCY_LOADING = 'SET_CURRENCY_LOADING';
export const SET_CURRENCY_REQUEST = 'SET_CURRENCY_REQUEST';
export const SET_CURRENCY_IDS = 'SET_CURRENCY_IDS';
export const RESET_CURRENCY = 'RESET_CURRENCY';

export const CURRENCY_API = `/currencies`;
