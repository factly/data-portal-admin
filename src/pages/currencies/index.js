import React from 'react';
import CurrenciesList from './components/CurrenciesList';
import { Button } from 'antd';
import { Link } from 'react-router-dom';

const Currencies = () => {
  return (
    <>
      <Link to={'/currencies/create'}>
        <Button type="primary" style={{ marginBottom: 16 }}>
          Add Currency
        </Button>
      </Link>
      <CurrenciesList />
    </>
  );
};

export default Currencies;
