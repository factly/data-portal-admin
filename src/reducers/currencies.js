import produce from 'immer';
import {
  ADD_CURRENCY,
  ADD_CURRENCIES,
  SET_CURRENCY_LOADING,
  SET_CURRENCY_REQUEST,
  SET_CURRENCY_IDS,
  RESET_CURRENCY,
} from '../constants/currencies';

const initialState = {
  loading: false,
  ids: [],
  req: [],
  items: {},
  total: 0,
};

const currenciesReducer = produce((draft, action = {}) => {
  switch (action.type) {
    case SET_CURRENCY_LOADING:
      draft.loading = action.payload.loading;
      return;
    case ADD_CURRENCY: {
      const { currency } = action.payload;
      draft.items[currency.id] = currency;
      return;
    }
    case ADD_CURRENCIES: {
      const { currencies } = action.payload;
      Object.assign(draft.items, currencies);
      return;
    }
    case SET_CURRENCY_IDS:
      draft.ids = action.payload.ids;
      return;
    case SET_CURRENCY_REQUEST: {
      const { req, total } = action.payload;
      draft.req.push(req);
      draft.total = total;
      return;
    }
    case RESET_CURRENCY:
      return initialState;
  }
}, initialState);

export default currenciesReducer;
