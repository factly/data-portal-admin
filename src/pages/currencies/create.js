import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form, Input, Button, notification } from 'antd';

import { createCurrency } from '../../actions/currencies';

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const CurrencyCreate = (props) => {
  const { create } = props;

  const onFinish = (values) => {
    create(values)
      .then(() => {
        notification.success({
          message: 'Success',
          description: 'Currency succesfully added',
        });
        props.history.push('/currencies');
      })
      .catch(() => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  return (
    <Form name="currencies_create" {...formItemLayout} onFinish={onFinish}>
      <Form.Item
        label="Name"
        name="name"
        rules={[
          {
            required: true,
            message: 'Please enter name!',
          },
        ]}
      >
        <Input placeholder="Ex. Indian Rupee" />
      </Form.Item>

      <Form.Item
        label="ISO Code"
        name="iso_code"
        rules={[
          {
            required: true,
            message: 'Please enter iso code!',
          },
        ]}
      >
        <Input placeholder="Ex. INR" />
      </Form.Item>

      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

CurrencyCreate.propTypes = {
  create: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  create: (values) => dispatch(createCurrency(values)),
});

export default connect(null, mapDispatchToProps)(CurrencyCreate);
