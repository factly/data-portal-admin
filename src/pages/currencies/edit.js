import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';

import CurrencyCreateForm from './components/CurrencyCreateForm';
import { getCurrency, updateCurrency } from '../../actions/currencies';

const EditCurrency = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const { currency } = useSelector(({ currencies }) => currencies);

  React.useEffect(() => {
    dispatch(getCurrency(id));
  }, []);

  if (!currency.id) return <Skeleton />;

  const onUpdate = async (values) => {
    await dispatch(updateCurrency(id, values));
    history.push('/currencies');
  };

  return <CurrencyCreateForm data={currency} onSubmit={onUpdate} />;
};

export default EditCurrency;
