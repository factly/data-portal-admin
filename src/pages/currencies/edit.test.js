import React from 'react';
import { useHistory } from 'react-router-dom';
import renderer, { act as rendererAct } from 'react-test-renderer';
import { useDispatch, Provider, useSelector } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import EditCurrency from './edit';
import * as actions from '../../actions/currencies';
import CurrencyEditForm from './components/CurrencyCreateForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
  useParams: jest.fn().mockReturnValue({ id: '1' }),
}));

jest.mock('../../actions/currencies', () => ({
  getCurrency: jest.fn(),
  addCurrency: jest.fn(),
  updateCurrency: jest.fn(),
}));

describe('Currencies List component', () => {
  const store = mockStore({
    currencies: {
      req: [],
      details: {
        1: {
          id: 1,
          name: 'Currency-1',
          slug: 'currency-1',
          description: 'description',
        },
        2: {
          id: 2,
          name: 'Currency-2',
          slug: 'currency-2',
          description: 'description',
        },
      },
      loading: true,
    },
  });
  store.dispatch = jest.fn(() => ({}));

  const mockedDispatch = jest.fn();
  useDispatch.mockReturnValue(mockedDispatch);

  describe('snapshot testing', () => {
    it('should render the component', () => {
      useSelector.mockReturnValueOnce({
        currency: {
          id: 1,
          name: 'currency',
          slug: 'slug',
          description: 'description',
        },
        loading: false,
      });
      let component;
      rendererAct(() => {
        component = renderer.create(
          <Provider store={store}>
            <EditCurrency />
          </Provider>,
        );
      });
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should match component with empty data', () => {
      useSelector.mockReturnValueOnce({
        currency: {},
        loading: false,
      });
      let component;
      store.details = {};
      rendererAct(() => {
        component = renderer.create(
          <Provider store={store}>
            <EditCurrency />
          </Provider>,
        );
      });
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should match skeleton while loading', () => {
      useSelector.mockReturnValueOnce({
        currency: {},
        loading: true,
      });
      let component;
      rendererAct(() => {
        component = renderer.create(
          <Provider store={store}>
            <EditCurrency />
          </Provider>,
        );
      });
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call get action', () => {
      useSelector.mockReturnValueOnce({ currency: null, loading: true });
      actions.getCurrency.mockReset();
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditCurrency />
          </Provider>,
        );
      });
      expect(actions.getCurrency).toHaveBeenCalledWith('1');
    });
    it('should call updateCurrency', () => {
      useSelector.mockReturnValueOnce({ currency: {}, loading: false });
      actions.updateCurrency.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditCurrency />
          </Provider>,
        );
      });
      wrapper.find(CurrencyEditForm).props().onSubmit({ test: 'test' });
      expect(actions.updateCurrency).toHaveBeenCalledWith('1', { test: 'test' });
      expect(push).toHaveBeenCalledWith('/currencies');
    });
  });
});
