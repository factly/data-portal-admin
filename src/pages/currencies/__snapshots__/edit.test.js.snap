// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currencies List component snapshot testing should match component with empty data 1`] = `
<form
  className="ant-form ant-form-horizontal"
  id="currencies_create"
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-6 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="currencies_create_name"
        title="Name"
      >
        Name
      </label>
    </div>
    <div
      className="ant-col ant-col-14 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="currencies_create_name"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Ex. Indian Rupee"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-6 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="currencies_create_iso_code"
        title="ISO Code"
      >
        ISO Code
      </label>
    </div>
    <div
      className="ant-col ant-col-14 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="currencies_create_iso_code"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Ex. INR"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-12 ant-col-offset-6 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <button
            className="ant-btn ant-btn-primary"
            onClick={[Function]}
            type="submit"
          >
            <span>
              Submit
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`Currencies List component snapshot testing should match skeleton while loading 1`] = `
<div
  className="ant-skeleton"
>
  <div
    className="ant-skeleton-content"
  >
    <h3
      className="ant-skeleton-title"
      style={
        Object {
          "width": "38%",
        }
      }
    />
    <ul
      className="ant-skeleton-paragraph"
    >
      <li
        style={
          Object {
            "width": undefined,
          }
        }
      />
      <li
        style={
          Object {
            "width": undefined,
          }
        }
      />
      <li
        style={
          Object {
            "width": "61%",
          }
        }
      />
    </ul>
  </div>
</div>
`;

exports[`Currencies List component snapshot testing should render the component 1`] = `
<form
  className="ant-form ant-form-horizontal"
  id="currencies_create"
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-6 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="currencies_create_name"
        title="Name"
      >
        Name
      </label>
    </div>
    <div
      className="ant-col ant-col-14 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="currencies_create_name"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Ex. Indian Rupee"
            type="text"
            value="currency"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-6 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="currencies_create_iso_code"
        title="ISO Code"
      >
        ISO Code
      </label>
    </div>
    <div
      className="ant-col ant-col-14 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="currencies_create_iso_code"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Ex. INR"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-12 ant-col-offset-6 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <button
            className="ant-btn ant-btn-primary"
            onClick={[Function]}
            type="submit"
          >
            <span>
              Submit
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;
